files/wgsyncer.py

```diff
@@ -0,0 +1,373 @@
+                old_allowed_ips, prefix, parse_ip(attrs["RTA_GATEWAY"]))
+        for allowed_ip in generator:
+            new_allowed_ips_dict[allowed_ip[0]].append(allowed_ip[1])
+
+    def filtered_prefix_list(prefix_list):
+        def should_keep(prefix):
+            return all(map(lambda other_prefix: not Prefix.is_strict_subset(
+                prefix, other_prefix), prefix_list))
+        return list(filter(should_keep, prefix_list))
+    return {pubkey: filtered_prefix_list(
+        prefix_list) for pubkey, prefix_list in new_allowed_ips_dict.items()}
+
+
+def parse_ip(ip_str):
+    """Converts an ip from str to int."""
+    split = ip_str.split(".")
+    ip_int = 0
+    for block in split:
+        ip_int *= 256
+        ip_int += int(block)
+    return ip_int
+
+
+def convert_link_route(old_allowed_ips, prefix):
+    """
+    Args:
+        old_allowed_ips (list): See documentation of calc_new_allowed_ips_dict.
+        prefix (Prefix): A prefix that is sent to the WireGuard interface for link-local routing.
+
+    Returns:
+        A generator yielding the allowed IPs that have to be added s.t. the prefix can be
+        successfully routed to the WireGuard interface. Hereby each yielded element is a pair
+        (pubkey: str, prefix: Prefx).
+    """
+    for entry in old_allowed_ips:
+        if Prefix.is_subset(prefix, entry[1]):
+            yield (entry[0], prefix)
+            break
+        if Prefix.is_subset(entry[1], prefix):
+            yield entry
+
+
+def convert_global_route(old_allowed_ips, prefix, gateway):
+    """
+    Args:
+        old_allowed_ips (list): See documentation of calc_new_allowed_ips_dict.
+        prefix (Prefix):
+            A prefix that is sent to the WireGuard interface for global routing via some gateway.
+        gateway (int): The gateway described above, given as int.
+
+    Returns:
+        See documentation of convert_link_route.
+    """
+    for entry in old_allowed_ips:
+        if Prefix.is_subset(Prefix(gateway, 32), entry[1]):
+            yield (entry[0], prefix)
+            break
+
+
+def set_allowed_ips(interface_name, allowed_ips_dict):
+    """
+    Sets the allowed IPs for a WireGuard interface.
+
+    Args:
+        interface_name (str): Name of the WireGuard interface.
+        allowed_ips_dict (dict):
+            dict with pubkeys as key where each value is the respective list of allowed IPs (as
+            Prefix objects).
+    """
+    for pubkey, prefix_list in sorted(allowed_ips_dict.items()):
+        comma_separated_prefix_list = ",".join(map(str, prefix_list))
+        command = ["wg", "set", interface_name, "peer",
+                   pubkey, "allowed-ips", comma_separated_prefix_list]
+        logging.info("+ %s", " ".join(command))
+        try:
+            subprocess.run(command, check=True)
+        except subprocess.CalledProcessError as exc:
+            logging.error("subprocess exited with code %s.", exc.returncode)
```

Comment:
No:

* Then the script would fail if only one of multiple instances has a problem.
* Then the script would fail if an instance has a temporary problem, i.e. somebody deletes and re-adds the WireGuard interface.